src/Types.hs

```diff
@@ -2,35 +2,100 @@
 
 module Types (
   DiscoveredProject (..),
+  FoundTargets (..),
   BuildTarget (..),
   LicenseResult (..),
   License (..),
   LicenseType (..),
   module DepTypes,
+  TargetFilter (..),
 ) where
 
-import Data.Aeson
-import Data.Set (Set)
+import Data.Aeson (
+  FromJSON (parseJSON),
+  KeyValue ((.=)),
+  ToJSON (toJSON),
+  object,
+  withObject,
+  (.:),
+  (.:?),
+ )
+
+import Data.Aeson.Types (Parser)
+import Data.Set qualified as S
+import Data.Set.NonEmpty (NonEmptySet, nonEmpty, toSet)
 import Data.Text (Text)
-import DepTypes
-import Graphing
-import Path
+import Data.Text qualified as T
+import DepTypes (
+  DepEnvironment (..),
+  DepType (..),
+  Dependency (..),
+  VerConstraint (..),
+  insertEnvironment,
+  insertLocation,
+  insertTag,
+ )
+import Graphing (Graphing)
+import Path (Abs, Dir, Path, Rel, parseRelDir)
 
 -- TODO: results should be within a graph of build targets && eliminate SubprojectType
 
+data FoundTargets = ProjectWithoutTargets | FoundTargets (NonEmptySet BuildTarget)
+  deriving (Eq, Ord, Show)
+
+instance Semigroup FoundTargets where
+  (FoundTargets a) <> ProjectWithoutTargets = FoundTargets a
+  ProjectWithoutTargets <> (FoundTargets a) = FoundTargets a
+  (FoundTargets a) <> (FoundTargets b) = maybe ProjectWithoutTargets FoundTargets (nonEmpty (S.union (toSet a) (toSet b)))
+  ProjectWithoutTargets <> ProjectWithoutTargets = ProjectWithoutTargets
```

Comment:
[nit] this is great and covers all of the cases; we can simplify this a bit if we squint at some of these

specifically, it looks like `ProjectWithoutTargets <> anything` is `anything`:

```hs
  ProjectWithoutTargets <> a = a
  a <> ProjectWithoutTargets = a
  (FoundTargets a) <> (FoundTargets b) = maybe ProjectWithoutTargets FoundTargets (nonEmpty (S.union (toSet a) (toSet b)))
```

additionally, if we define a `Semigroup` instance on `NonEmptySet` (which I think we can throw in the `deriving` clause -- because `<>` on `Set a` is `Set.union`):

```suggestion
  ProjectWithoutTargets <> a = a
  a <> ProjectWithoutTargets = a
  FoundTargets a <> FoundTargets b = FoundTargets (a <> b)
```

This approximately mirrors the `Semigroup a => Semigroup (Maybe a)` instance as well:
```hs
instance Semigroup a => Semigroup (Maybe a) where
  Nothing <> a = a
  a <> Nothing = a
  Just a <> Just b = Just (a <> b)
```

..because `FoundTargets` is isomorphic to `Maybe (NonEmptySet BuildTarget)`